Add `cg.SycamoreTargetGateset` and deprecate `cg.ConvertToSycamoreGates`

tanujkhattar opened the pull request:
- Part of https://github.com/quantumlib/Cirq/issues/5028
- Follows the new Transformer API https://github.com/quantumlib/Cirq/issues/4483
- Supports no compile tags NoCompile Tag for optimizers https://github.com/quantumlib/Cirq/issues/4253

verult commented on cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py:
```diff
@@ -0,0 +1,153 @@
+        circuit,
+        merge_func_swap_rzz,
+        tags_to_ignore=tags_to_ignore,
+        merged_circuit_op_tag=merged_swap_rzz_tag,
```
Why do we tag these connected components? Is it for users to read the circuit and identify transformation steps?

tanujkhattar replied:
Yes, the tags are useful in logging but also useful for subsequent transformation steps where we want to identify whether a newly created connected component is a swap + rzz component (and hence is a known decomposition) or is an arbitrary 2q connected component and should be decomposed analytically or using the tabulation.

verult commented on cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py:
```diff
@@ -0,0 +1,351 @@
+    circuit1 = cirq.Circuit(
+        cirq.SWAP(qubits[0], qubits[1]),
+        cirq.Z(qubits[2]),
+        cirq.ZZ(qubits[0], qubits[1]) ** gamma,
```
Does differing values of `gamma` change the the compiled circuit structure drastically? Wondering why you chose not to assert the compiled circuit against a hard-coded circuit.

tanujkhattar replied:
The tests were copied from previous implementation of convert_to_sycamore_gates.py and I don't think changing gamma to a hardcoded value would make much of a difference. 

The important thing to test is that swap + rzz does get compiled via the known compilations flow.

verult commented on cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py:
```diff
@@ -0,0 +1,351 @@
+        )
+
+
+def test_unsupported_gate_ignoring_failures():
```
nit: Add comment explaining why it's OK to ignore the failure.

tanujkhattar replied:
Ignoring failures can be useful in a variety of different use cases. For example, when this method occurs is called as part of a larger transformer pipeline and the user has subsequent transformation steps to convert any unsupported gates. 

I think the flag name + args are clear enough to describe what they do and answering the "why" is beyond the scope here.

verult replied:
Ah my confusion was around how "ignoring failures" is being tested, but realized later that `ignore_failures` defaults to True in `optimize_for_target_gateset()` :) Explicitly setting the parameter in the call here would be helpful in reading the test case, but again very much a nit.

verult reviewed (commented):
lgtm other than some minor comments

tanujkhattar commented on cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py:
```diff
@@ -0,0 +1,153 @@
+            cirq.ZPowGate,
+            name='SycamoreTargetGateset',
+        )
+        if tabulation is not None and not isinstance(tabulation, cirq.TwoQubitGateTabulation):
```
Removed the assertion, it got copied over from the previous implementation that had a similar flow.

tanujkhattar commented on cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py:
```diff
@@ -0,0 +1,153 @@
+
+        Args:
+            atol: A limit on the amount of absolute error introduced by the decomposition.
+            tabulation: If set, a tabulation for the Sycamore gate to use for decomposing Matrix
```
Reworded the docstring to reflect this better.

tanujkhattar commented:
@verult Addressed all your comments. Please let me know if you have any further feedback or else approve the PR so I can submit. Thanks for the review!
